jerry-ext/handle-scope/handle-scope.c

```diff
@@ -0,0 +1,370 @@
+{
+  return jerryx_close_handle_scope (scope);
+} /** jerryx_close_escapable_handle_scope */
+
+
+/**
+ * Internal helper.
+ * Escape a jerry value from the scope, yet did not promote it to outer scope.
+ * An assertion of if parent exists shall be made before invoking this function.
+ *
+ * @param scope - scope of the handle added to.
+ * @param idx - expected index of the handle in the scope's prelist.
+ * @returns escaped jerry value id
+ */
+jerry_value_t
+jerryx_hand_scope_escape_handle_from_prelist (jerryx_handle_scope scope, size_t idx)
+{
+  jerry_value_t jval = scope->handle_prelist[idx];
+  if (scope->handle_count > JERRYX_HANDLE_PRELIST_SIZE)
+  {
+    jerryx_handle_t *handle = scope->handle_ptr;
+    scope->handle_ptr = handle->sibling;
+    scope->handle_prelist[idx] = handle->jval;
+    return jval;
+  }
+
+  if (idx < JERRYX_HANDLE_PRELIST_SIZE - 1)
+  {
+    scope->handle_prelist[idx] = scope->handle_prelist[scope->handle_count - 1];
+  }
+  return jval;
+} /** jerryx_hand_scope_escape_handle_from_prelist */
+
+
+/**
+ * Internal helper.
+ * Escape a jerry value from the given escapable handle scope.
+ *
+ * @param scope - the expected scope to be escaped from.
+ * @param escapee - the jerry value to be escaped.
+ * @param result - [out value] escaped jerry value result.
+ * @param should_promote - true if the escaped value should be added to parent
+ *                         scope after escaped from the given handle scope.
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_escape_handle_internal (jerryx_escapable_handle_scope scope,
+                               jerry_value_t escapee,
+                               jerry_value_t *result,
+                               bool should_promote)
+{
+  if (scope->escaped)
+  {
+    return jerryx_escape_called_twice;
+  }
+
+  jerryx_handle_scope parent = jerryx_handle_scope_get_parent (scope);
+  if (parent == NULL)
+  {
+    return jerryx_handle_scope_mismatch;
+  }
+
+  bool found = false;
+  {
+    size_t found_idx = 0;
+    size_t prelist_count = (scope->handle_count < JERRYX_HANDLE_PRELIST_SIZE ? scope->handle_count
+                                                                             : JERRYX_HANDLE_PRELIST_SIZE);
+    /**
+     * Search prelist in a reversed order since last added handle
+     * is possible the one to be escaped
+     */
+    for (size_t idx_plus_1 = prelist_count; idx_plus_1 > 0; --idx_plus_1)
+    {
+      if (escapee == scope->handle_prelist[idx_plus_1 - 1])
+      {
+        found = true;
+        found_idx = idx_plus_1 - 1;
+        break;
+      }
+    }
+
+    if (found)
+    {
+      *result = jerryx_hand_scope_escape_handle_from_prelist (scope, found_idx);
+      if (should_promote)
+      {
+        /**
+         * Escape handle to parent scope
+         */
+        jerryx_create_handle_in_scope (*result, jerryx_handle_scope_get_parent (scope));
+      }
+      goto deferred;
```

Comment:
I think this goto is unnecessary. The code snippet where it jumps can be extracted here directly.